lib/routes/utils/signin.js

```diff
@@ -0,0 +1,485 @@
+
+const BASE_36 = validators.BASE_36
+
+// An arbitrary, but very generous, limit on the number of active sessions.
+// Currently only for metrics purposes, not enforced.
+const MAX_ACTIVE_SESSIONS = 200
+
+module.exports = (log, config, customs, db, mailer)  => {
+
+  const unblockCodeLifetime = config.signinUnblock && config.signinUnblock.codeLifetime || 0
+  const unblockCodeLen = config.signinUnblock && config.signinUnblock.codeLength || 8
+
+  return {
+
+    validators: {
+      UNBLOCK_CODE: isA.string().regex(BASE_36).length(unblockCodeLen).optional()
+    },
+
+    /**
+     * Check if the password a user entered matches the one on
+     * file for the account. If it does not, flag the account with
+     * customs. Higher level code will take care of
+     * returning an error to the user.
+     */
+    checkPassword(accountRecord, password, clientAddress) {
+      if (butil.buffersAreEqual(accountRecord.authSalt, butil.ONES)) {
+        return customs.flag(clientAddress, {
+          email: accountRecord.email,
+          errno: error.ERRNO.ACCOUNT_RESET
+        }).then(() => {
+          throw error.mustResetAccount(accountRecord.email)
+        })
+      }
+      return password.verifyHash()
+        .then(verifyHash => {
+          return db.checkPassword(accountRecord.uid, verifyHash)
+        })
+        .then(match => {
+          if (match) {
+            return match
+          }
+          return customs.flag(clientAddress, {
+            email: accountRecord.email,
+            errno: error.ERRNO.INCORRECT_PASSWORD
+          }).then(() => {
+            match
+          })
+        })
+    },
+
+    /**
+     * Check if the user is logging in with the correct email address
+     * for their account.  If not, throw an error that will help them
+     * find the right one.
+     */
+    checkEmailAddress(accountRecord, email, originalLoginEmail) {
+      // The `originalLoginEmail` param is specified in the login request if the user has
+      // changed their primary email address, and hence needed to use a different email for
+      // calculating authPW than they use for identifying their account.
+      if (originalLoginEmail) {
+        // They must have been originally trying to log in with the correct
+        // primary email address, not a secondary one.
+        if (originalLoginEmail.toLowerCase() !== accountRecord.primaryEmail.normalizedEmail) {
+          throw error.cannotLoginWithSecondaryEmail()
+        }
+
+        // If a user changes their primary email and then deletes the old primary
+        // email from their account, that email address is basically "bricked" -
+        // the user can't log in with it, and nobody can create a new account with it.
+        // The below causes us to return a special error when trying to sign in
+        // with such a "bricked" email, because otherwise the default error message
+        // would offer the user to create an account with that email, which they can't.
```

Comment:
I re-worded these comments in the process of trying to understand what this logic was for; please let me know if I've misunderstood!